File: backend/app/services/channel_manager.py
"""Channel management utilities."""

import uuid
from datetime import UTC, datetime

from sqlalchemy import select

from backend.app.db import async_session
from backend.app.models.agent import Agent
from backend.app.models.channel import Channel
from backend.app.models.channel_member import ChannelMember


async def list_all_channels() -> list[dict]:
    """List all channels."""
    async with async_session() as db:
        result = await db.execute(select(Channel).order_by(Channel.name))
        return [
            {
                "id": ch.id,
                "name": ch.name,
                "type": ch.type,
                "project_path": ch.project_path,
            }
            for ch in result.scalars().all()
        ]


async def join_agent_to_channel(agent_name: str, channel_name: str) -> dict:
    """Add an agent to a channel."""
    now = datetime.now(UTC).isoformat()

    async with async_session() as db:
        agent_result = await db.execute(select(Agent).where(Agent.agent_name == agent_name))
        agent = agent_result.scalar_one_or_none()
        if not agent:
            return {"error": "Agent not found."}

        ch_result = await db.execute(select(Channel).where(Channel.name == channel_name))
        ch = ch_result.scalar_one_or_none()
        if not ch:
            return {"error": f"Channel '{channel_name}' not found."}

        # Check if already member
        existing = await db.execute(
            select(ChannelMember).where(
                ChannelMember.channel_id == ch.id,
                ChannelMember.user_id == agent.id,
            )
        )
        if existing.scalar_one_or_none():
            return {"status": "already_member"}

        db.add(ChannelMember(channel_id=ch.id, user_id=agent.id, joined_at=now))
        await db.commit()

    return {"status": "joined", "channel": channel_name}


async def create_new_channel(name: str, created_by: str) -> dict:
    """Create a new channel."""
    channel_name = name if name.startswith("#") else f"#{name}"

    async with async_session() as db:
        existing = await db.execute(select(Channel).where(Channel.name == channel_name))
        if existing.scalar_one_or_none():
            return {"error": f"Channel '{channel_name}' already exists."}

        channel = Channel(
            id=str(uuid.uuid4()),
            name=channel_name,
            type="custom",
            created_by=created_by,
            created_at=datetime.now(UTC).isoformat(),
        )
        db.add(channel)
        await db.commit()

        return {"id": channel.id, "name": channel.name, "type": channel.type}

File: server/src/services/channel-manager.ts
/**
 * Channel management utilities — used by MCP tools and REST API.
 */

import { eq, and } from "drizzle-orm";
import { getDb } from "../db";
import { agents, channels, channelMembers } from "../db/schema";
import { broadcastEvent, channelCreatedEvent } from "./broadcaster";

export function listAllChannels(): Array<{
  id: string;
  name: string;
  type: string;
  topic: string | null;
  project_path: string | null;
}> {
  const db = getDb();
  return db
    .select()
    .from(channels)
    .orderBy(channels.name)
    .all()
    .map((ch) => ({
      id: ch.id,
      name: ch.name,
      type: ch.type,
      topic: ch.topic,
      project_path: ch.projectPath,
    }));
}

export function joinAgentToChannel(
  agentName: string,
  channelName: string
): Record<string, unknown> {
  const db = getDb();
  const now = new Date().toISOString();

  const agent = db
    .select()
    .from(agents)
    .where(eq(agents.agentName, agentName))
    .get();
  if (!agent) return { error: "Agent not found." };

  const ch = db.select().from(channels).where(eq(channels.name, channelName)).get();
  if (!ch) return { error: `Channel '${channelName}' not found.` };

  // Check if already member
  const existing = db
    .select()
    .from(channelMembers)
    .where(
      and(
        eq(channelMembers.channelId, ch.id),
        eq(channelMembers.userId, agent.id)
      )
    )
    .get();
  if (existing) return { status: "already_member" };

  db.insert(channelMembers)
    .values({ channelId: ch.id, userId: agent.id, joinedAt: now })
    .run();

  return { status: "joined", channel: channelName };
}

export function createNewChannel(
  name: string,
  createdBy: string
): Record<string, unknown> {
  const channelName = name.startsWith("#") ? name : `#${name}`;
  const db = getDb();

  const existing = db
    .select()
    .from(channels)
    .where(eq(channels.name, channelName))
    .get();
  if (existing) return { error: `Channel '${channelName}' already exists.` };

  const id = crypto.randomUUID();
  const now = new Date().toISOString();

  db.insert(channels)
    .values({
      id,
      name: channelName,
      type: "custom",
      createdBy,
      createdAt: now,
    })
    .run();

  broadcastEvent(
    channelCreatedEvent({
      channelId: id,
      channelName,
      channelType: "custom",
    })
  );

  return { id, name: channelName, type: "custom" };
}

export function setChannelTopic(
  channelName: string,
  topic: string
): Record<string, unknown> {
  const db = getDb();
  const ch = db.select().from(channels).where(eq(channels.name, channelName)).get();
  if (!ch) return { error: `Channel '${channelName}' not found.` };

  db.update(channels)
    .set({ topic: topic || null })
    .where(eq(channels.id, ch.id))
    .run();

  return { status: "updated", channel: channelName, topic: topic || null };
}
